File: py/src/aoc2020/day22.py
from collections import deque
import fileinput


def parse(lines):
    it = iter(lines)
    line = next(it)
    assert line.rstrip() == 'Player 1:'
    deck1 = deque()
    while (line := next(it)).rstrip():
        deck1.append(int(line))
    line = next(it)
    assert line.rstrip() == 'Player 2:'
    deck2 = deque()
    while True:
        try:
            line = next(it)
        except StopIteration:
            break
        deck2.append(int(line))
    return deck1, deck2


def part1(lines):
    '''
    >>> part1(('Player 1:', '9', '2', '6', '3', '1', '',
    ...        'Player 2:', '5', '8', '4', '7', '10'))
    306
    '''
    deck1, deck2 = parse(lines)
    while deck1 and deck2:
        card1 = deck1.popleft()
        card2 = deck2.popleft()
        if card1 < card2:
            deck2.append(card2)
            deck2.append(card1)
        else:
            deck1.append(card1)
            deck1.append(card2)
    winner = deck1 or deck2
    return sum((len(winner) - i) * x for i, x in enumerate(winner))


def part2(lines):
    '''
    >>> part2(('Player 1:', '9', '2', '6', '3', '1', '',
    ...        'Player 2:', '5', '8', '4', '7', '10'))
    291
    '''
    deck1, deck2 = parse(lines)
    go(deck1, deck2)
    winner = deck1 or deck2
    return sum((len(winner) - i) * x for i, x in enumerate(winner))


def go(deck1, deck2):
    seen = set()
    while deck1 and deck2:
        state = tuple(deck1), tuple(deck2)
        if state in seen:
            return False
        seen.add(state)
        card1 = deck1.popleft()
        card2 = deck2.popleft()
        if (go(deque(list(deck1)[:card1]), deque(list(deck2)[:card2]))
                if card1 <= len(deck1) and card2 <= len(deck2) else
                card1 < card2):
            deck2.append(card2)
            deck2.append(card1)
        else:
            deck1.append(card1)
            deck1.append(card2)
    return not deck1


parts = (part1, part2)

if __name__ == '__main__':
    lines = list(fileinput.input())
    print(part1(lines))
    print(part2(lines))

File: rs/src/day22.rs
use super::util;
use std::collections::{HashSet, VecDeque};
use std::error::Error;

fn parse<'a, I, S>(
    lines: I,
) -> Result<(VecDeque<usize>, VecDeque<usize>), Box<dyn Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let mut iter = lines.into_iter();
    if iter.next().map(|s| s.as_ref()) != Some("Player 1:") {
        return Err(util::Error.into());
    }
    let mut deck = VecDeque::new();
    while let Some(line) = iter.next() {
        let line = line.as_ref();
        if line.is_empty() {
            break;
        }
        deck.push_back(line.parse()?);
    }
    if iter.next().map(|s| s.as_ref()) != Some("Player 2:") {
        return Err(util::Error.into());
    }
    Ok((
        deck,
        iter.map(|s| s.as_ref())
            .map(|s| s.parse())
            .collect::<Result<_, _>>()?,
    ))
}

pub fn part1<'a, I, S>(lines: I) -> Result<usize, Box<dyn Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let (mut deck1, mut deck2) = parse(lines)?;
    while !deck1.is_empty() && !deck2.is_empty() {
        let card1 = deck1.pop_front().ok_or(util::Error)?;
        let card2 = deck2.pop_front().ok_or(util::Error)?;
        if card1 < card2 {
            deck2.push_back(card2);
            deck2.push_back(card1);
        } else {
            deck1.push_back(card1);
            deck1.push_back(card2);
        }
    }
    let winner = Some(deck1).filter(|deck| !deck.is_empty()).unwrap_or(deck2);
    let n = winner.len();
    Ok(winner
        .into_iter()
        .enumerate()
        .map(|(i, x)| (n - i) * x)
        .sum())
}

pub fn part2<'a, I, S>(lines: I) -> Result<usize, Box<dyn Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let (mut deck1, mut deck2) = parse(lines)?;
    go(&mut deck1, &mut deck2).ok_or(util::Error)?;
    let winner = Some(deck1).filter(|deck| !deck.is_empty()).unwrap_or(deck2);
    let n = winner.len();
    Ok(winner
        .into_iter()
        .enumerate()
        .map(|(i, x)| (n - i) * x)
        .sum())
}

fn go(deck1: &mut VecDeque<usize>, deck2: &mut VecDeque<usize>) -> Option<bool> {
    let mut seen = HashSet::<(Vec<_>, Vec<_>)>::new();
    while !deck1.is_empty() && !deck2.is_empty() {
        if !seen.insert((
            deck1.iter().copied().collect(),
            deck2.iter().copied().collect(),
        )) {
            return Some(false);
        }
        let card1 = deck1.pop_front()?;
        let card2 = deck2.pop_front()?;
        let is_lt = if card1 <= deck1.len() && card2 <= deck2.len() {
            go(
                &mut deck1.iter().take(card1).copied().collect(),
                &mut deck2.iter().take(card2).copied().collect(),
            )?
        } else {
            card1 < card2
        };
        if is_lt {
            deck2.push_back(card2);
            deck2.push_back(card1);
        } else {
            deck1.push_back(card1);
            deck1.push_back(card2);
        }
    }
    Some(deck1.is_empty())
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &[&str] = &[
        "Player 1:",
        "9",
        "2",
        "6",
        "3",
        "1",
        "",
        "Player 2:",
        "5",
        "8",
        "4",
        "7",
        "10",
    ];

    #[test]
    fn part1_examples() -> Result<(), Box<dyn Error + Send + Sync>> {
        assert_eq!(306, part1(EXAMPLE)?);
        Ok(())
    }

    #[test]
    fn part2_examples() -> Result<(), Box<dyn Error + Send + Sync>> {
        assert_eq!(291, part2(EXAMPLE)?);
        Ok(())
    }
}
